sql/src/main/java/io/crate/protocols/postgres/BulkPortal.java

```diff
@@ -0,0 +1,197 @@
+import io.crate.executor.Executor;
+import io.crate.executor.TaskResult;
+import io.crate.operation.collect.StatsTables;
+import io.crate.planner.Plan;
+import io.crate.planner.Planner;
+import io.crate.sql.tree.Statement;
+import io.crate.types.DataType;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+import static io.crate.action.sql.SQLBulkRequest.EMPTY_BULK_ARGS;
+import static io.crate.action.sql.SQLRequest.EMPTY_ARGS;
+
+class BulkPortal extends AbstractPortal {
+
+    private String query;
+    private Statement statement;
+    private Analysis analysis;
+    private Plan plan;
+    private int maxRows = 0;
+    private List<? extends DataType> outputTypes;
+    private final List<List<Object>> bulkParams = new ArrayList<>();
+    private final List<ResultReceiver> resultReceivers = new ArrayList<>();
+
+    BulkPortal(String name,
+               String query,
+               Statement statement,
+               Analysis analysis,
+               Plan plan,
+               List<? extends DataType> outputTypes,
+               ResultReceiver resultReceiver,
+               int maxRows,
+               List<Object> params,
+               SessionData sessionData) {
+        super(name, sessionData);
+        this.query = query;
+        this.statement = statement;
+        this.analysis = analysis;
+        this.plan = plan;
+        this.outputTypes = outputTypes;
+        this.resultReceivers.add(resultReceiver);
+        this.maxRows = maxRows;
+        this.bulkParams.add(params);
+    }
+
+    @Override
+    public Type type() {
+        return Type.BULK;
+    }
+
+    @Override
+    public FormatCodes.FormatCode[] getResultFormatCodes() {
+        return new FormatCodes.FormatCode[0];
+    }
+
+    @Override
+    public List<? extends DataType> getOutputTypes() {
+        return outputTypes;
+    }
+
+    @Override
+    public String getLastQuery() {
+        return this.query;
+    }
+
+    @Override
+    public Portal bind(String statementName, String query, Statement statement,
+                List<Object> params, @Nullable FormatCodes.FormatCode[] resultFormatCodes) {
+        this.query = query;
+        this.statement = statement;
+        this.bulkParams.add(params);
+        if (analysis == null) {
+            analysis = sessionData.getAnalyzer().analyze(statement,
+                                                         new ParameterContext(getArgs(),
+                                                                              EMPTY_BULK_ARGS,
+                                                                              sessionData.getDefaultSchema()));
+        }
+        return this;
+    }
+
+    @Override
+    public List<Field> describe() {
+        if (analysis.rootRelation() == null) {
+            return null;
+        }
+        List<Field> fields = analysis.rootRelation().fields();
```

Comment:
does the bulkPortal support operations where there are rootRelations? I think this case here never happens